src/iotjs_string.c

```diff
@@ -98,57 +84,41 @@ bool iotjs_string_is_empty(const iotjs_string_t* str) {
 void iotjs_string_make_empty(iotjs_string_t* str) {
   IOTJS_VALIDATED_STRUCT_METHOD(iotjs_string_t, str);
 
-  if (_this->data != iotjs_empty_string) {
+  if (_this->data != NULL) {
     iotjs_buffer_release(_this->data);
-    _this->cap = _this->size = 0;
-    _this->data = iotjs_empty_string;
+    _this->size = 0;
+    _this->data = NULL;
   }
 }
 
 
-void iotjs_string_append(iotjs_string_t* str, const char* data, int size) {
+void iotjs_string_append(iotjs_string_t* str, const char* data, uint32_t size) {
   IOTJS_VALIDATED_STRUCT_METHOD(iotjs_string_t, str);
 
-  const int kCapacityIncreaseFactor = 2;
-
-  IOTJS_ASSERT(_this->data != NULL);
-  IOTJS_ASSERT(_this->cap >= _this->size);
+  IOTJS_ASSERT(data != NULL);
 
-  if (size < 0) {
-    size = strlen(data);
-  }
   if (size == 0) {
     return;
   }
 
-  if (_this->cap == 0) {
-    // No buffer was allocated.
-    IOTJS_ASSERT(_this->size == 0);
-    IOTJS_ASSERT(_this->data == iotjs_empty_string);
-    _this->cap = _this->size = size;
-    _this->data = iotjs_buffer_allocate(_this->cap + 1);
-    strncpy(_this->data, data, _this->size);
-  } else if (_this->cap >= _this->size + size) {
-    // Have enough capacity to append data.
-    strncpy(_this->data + _this->size, data, size);
-    _this->size += size;
+  if (_this->data != NULL) {
+    _this->data = iotjs_buffer_reallocate(_this->data, _this->size + size);
   } else {
-    // Lack of capacity, calculate next capacity.
-    while (_this->cap < _this->size + size) {
-      _this->cap *= kCapacityIncreaseFactor;
-    }
-    // Reallocate buffer and copy data.
-    _this->data = iotjs_buffer_reallocate(_this->data, _this->cap + 1);
-    strncpy(_this->data + _this->size, data, size);
-    _this->size += size;
+    IOTJS_ASSERT(_this->size == 0);
+    _this->data = iotjs_buffer_allocate(size);
   }
-  _this->data[_this->size] = 0;
+
+  strncpy(_this->data + _this->size, data, size);
```

Comment:
As @yichoi commented, we must consider string that contains '\0' inside. As far as I know, strncpy stops copying before size when it encounter '\0'.

```
$ cat h.cc

#include <stdio.h>
#include <string.h>

int main(void)
{
  char dst[64] = { 0, };
  char src[7] = { 'H', 'e', 'l', 0, 'l', 'o', 0 };

  strncpy(dst, src, 7);
  for (int i = 0; i < 7; i++)
    printf("dst[%d] = %d\n", i, dst[i]);

  return 0;
}

$ g++ -o h h.cc
$ ./h
dst[0] = 72
dst[1] = 101
dst[2] = 108
dst[3] = 0
dst[4] = 0
dst[5] = 0
dst[6] = 0
```